qutip/stochastic.py

```diff
@@ -599,27 +763,116 @@ def smesolve(H, rho0, times, c_ops=[], sc_ops=[], e_ops=[],
+                m_ops += [c + c.dag(), -1j * c - c.dag()]
+            sops += [c / np.sqrt(2), -1j / np.sqrt(2) * c]
+        sso.m_ops = m_ops
+        if not isinstance(sso.dW_factors, list):
+            sso.dW_factors = [np.sqrt(2)] * len(sops)
+        elif len(sso.dW_factors) == len(sso.sc_ops):
+            dW_factors = []
+            for fact in sso.dW_factors:
+                dW_factors += [np.sqrt(2) * fact, np.sqrt(2) * fact]
+            sso.dW_factors = dW_factors
+        elif len(sso.dW_factors) != len(sops):
+            raise Exception("The len of dW_factors is not the same as sc_ops")
+    else:
+        raise Exception("The method must be one of homodyne or heterodyne")
+
+    LH = 1 - (sso.H * 1j * sso.dt)
+    sso.pp = spre(sso.H) * 0
+    sso.sops = []
+    sso.preops = []
+    sso.postops = []
+    sso.preops2 = []
+    sso.postops2 = []
+
+    def _prespostdag(op):
+        return spre(op) * spost(op.dag())
+
+    for op in sso.c_ops:
+        LH -= op.norm() * sso.dt * 0.5
+        sso.pp += op.apply(_prespostdag)._f_norm2() * sso.dt
+
+    for i, op in enumerate(sops):
+        LH -= op.norm() * sso.dt * 0.5
+        sso.sops += [(spre(op) + spost(op.dag())) * sso.dt]
+        sso.preops += [spre(op)]
+        sso.postops += [spost(op.dag())]
+        for op2 in sops[i:]:
+            sso.preops2 += [spre(op * op2)]
+            sso.postops2 += [spost(op.dag() * op2.dag())]
+
+    sso.ce_ops = [QobjEvo(spre(op)) for op in sso.e_ops]
+    sso.cm_ops = [QobjEvo(spre(op)) for op in sso.m_ops]
+    sso.preLH = spre(LH)
+    sso.postLH = spost(LH.dag())
+    sso.preLH.compile()
+    sso.postLH.compile()
+    sso.pp.compile()
+    [op.compile() for op in sso.sops]
+    [op.compile() for op in sso.preops]
+    [op.compile() for op in sso.postops]
+    [op.compile() for op in sso.preops2]
+    [op.compile() for op in sso.postops2]
+    [op.compile() for op in sso.cm_ops]
+    [op.compile() for op in sso.ce_ops]
+
+    sso.solver_obj = pmsme
+    sso.solver_name = "smesolve_" + sso.solver
+    res = _sesolve_generic(sso, sso.options, sso.progress_bar)
+
+    if e_ops_dict:
+        res.expect = {e: res.expect[n]
+                      for n, e in enumerate(e_ops_dict.keys())}
+
+    return res
+
+
+def photocurrentmesolve(H, rho0, times, c_ops=[], sc_ops=[], e_ops=[],
+                        _safe_mode=True, args={}, **kwargs):
     """
-    A stochastic (piecewse deterministic process) PDP solver for wavefunction
-    evolution. For most purposes, use :func:`qutip.mcsolve` instead for quantum
-    trajectory simulations.
+    Solve stochastic master equation using the photocurrent method.
 
     Parameters
     ----------
 
-    H : :class:`qutip.Qobj`
+    H : :class:`qutip.Qobj`, or time dependent system.
         System Hamiltonian.
+        Can depend on time, see StochasticSolverOptions help for format.
 
-    psi0 : :class:`qutip.Qobj`
-        Initial state vector (ket).
+    rho/psi : :class:`qutip.Qobj`
```

Comment:
```suggestion
    rho0 : :class:`qutip.Qobj`
```
when it could be ket or dm we typically use `rho0`